vm/src/org/mozartoz/truffle/runtime/IdentityPair.java

```diff
@@ -0,0 +1,42 @@
+package org.mozartoz.truffle.runtime;
+
+/**
+ * An identity pair class to be used in general unification and equality
+ * algorithms. Allows recognizing identical pairs of objects within collections.
+ */
+public class IdentityPair implements Comparable<IdentityPair> {
+	private final Object a, b;
+	private final int hashCode;
+
+	public IdentityPair(Object a, Object b) {
+		if (a.hashCode() < b.hashCode()) {
```

Comment:
Should this be the identity Hash code and also be a single call saved in a local?